### PR TITLE
BAH-2361 | Using local time zone to display appointment date and time

### PR DESCRIPTION
**Pull request by sivareddyp**
JIRA - [https://bahmni.atlassian.net/browse/BAH-2361](https://bahmni.atlassian.net/browse/BAH-2361)

image

### CODE REVIEW

**Comment by gsluthra** [comment]
Is it possible to add some unit tests which check this conversion logic for 2 timezones.. say India and Kenya timezones? Just to ensure it works fine?

**Comment by gsluthra** [comment]
If this is an existing control, has this bug always existed?

**Comment by sivareddyp** [comment]
> Is it possible to add some unit tests which check this conversion logic for 2 timezones.. say India and Kenya timezones? Just to ensure it works fine?

@gsluthra I dont think we can unit tests in the config. but, I will work with @gokultw to have unit tests in E2E test.

**Comment by sivareddyp** [comment]
> If this is an existing control, has this bug always existed?

@gsluthra clinic config squad took the reference from [default config](https://github.com/Bahmni/default-config/blob/master/openmrs/apps/customDisplayControl/js/customControl.js#L179),  and that is not considering timezone differences. Issue arises when the time zone for MySQL DB server and client time zone differs. [Demo Server's MySQL configured](https://demo.mybahmni.org/openmrs/admin/maintenance/systemInfo.htm) with Asia/Kolkata as Time Zone. Current [lab lite ](https://dev.lite.mybahmni.in/openmrs/admin/maintenance/systemInfo.htm) MySQL using UTC as time zone. 
image

File: openmrs/apps/customDisplayControl/js/customControl.js [hunk 1]
```diff
@@ -53,6 +62,21 @@ angular.module('bahmni.common.displaycontrol.custom')
...
+                let slotStartTime = $scope.pastAppointments[i].DASHBOARD_APPOINTMENTS_SLOT_KEY.split("-")[0].trim();
+                let slotEndTime = $scope.pastAppointments[i].DASHBOARD_APPOINTMENTS_SLOT_KEY.split("-")[1].trim();
+                let slotStartTimeInLocalTimeZone = getUpdatedDateTimeInLocalTimeZone(Bahmni.Common.Util.DateUtil.formatDateWithTime(appointmentDate + " " + slotStartTime));
+                let slotEndTimeInLocalTimeZone = getUpdatedDateTimeInLocalTimeZone(Bahmni.Common.Util.DateUtil.formatDateWithTime(appointmentDate + " " + slotEndTime));
+                let updatedSlotStartTime = Bahmni.Common.Util.DateUtil.formatTime(slotStartTimeInLocalTimeZone);
+                let updatedSlotEndTime = Bahmni.Common.Util.DateUtil.formatTime(slotEndTimeInLocalTimeZone);
+                $scope.pastAppointments[i].DASHBOARD_APPOINTMENTS_SLOT_KEY = updatedSlotStartTime + " - " + updatedSlotEndTime;
+                $scope.pastAppointments[i].DASHBOARD_APPOINTMENTS_DATE_KEY = Bahmni.Common.Util.DateUtil.formatDateWithoutTime(slotStartTimeInLocalTimeZone);
+            }
```
Comment (N0-man) [inline openmrs/apps/customDisplayControl/js/customControl.js:100]: a bit WET and hard to read... you are basically mutating appointment date and time post transformation

```
for (let i = 0; i < $scope.pastAppointments.length; i++) {
  const appointmentDate = 
    transformDate($scope.pastAppointments[i].DASHBOARD_APPOINTMENTS_DATE_KEY.split('/'))

  let slotStartTime = transformToLocalTime(
    appointmentDate,
    $scope.pastAppointments[i].DASHBOARD_APPOINTMENTS_SLOT_KEY.split('-')[0].trim()
  )
  
  let slotEndTime = transformToLocalTime(
    appointmentDate,
    $scope.pastAppointments[i].DASHBOARD_APPOINTMENTS_SLOT_KEY.split('-')[1].trim()
  )

  $scope.pastAppointments[i].DASHBOARD_APPOINTMENTS_SLOT_KEY =
    slotStartTime + ' - ' + slotEndTime
  $scope.pastAppointments[i].DASHBOARD_APPOINTMENTS_DATE_KEY =
    appointmentDate
}
```

and the functions are...
```
     var transformDate = function (appointmentDate) {
          let appointmentDay = appointmentDate[0]
          let appointmentMonth = appointmentDate[1]
          let appointmentYear = appointmentDate[2]
          return Bahmni.Common.Util.DateUtil.formatDateWithoutTime(
            new Date(appointmentYear, appointmentMonth - 1, appointmentDay),
          )
        }

        var transformToLocalTime = function (appointmentDate, slotTime) {
          let slotTimeInLocalTimeZone = getUpdatedDateTimeInLocalTimeZone(
            Bahmni.Common.Util.DateUtil.formatDateWithTime(
              appointmentDate + ' ' + slotTime,
            ),
          )
          return Bahmni.Common.Util.DateUtil.formatTime(slotTimeInLocalTimeZone)
        }
```

Comment (sivareddyp) [reply openmrs/apps/customDisplayControl/js/customControl.js:100]: @N0-man updated the PR.. thanks

**Review by N0-man — APPROVED** [review]

**Pull request merged**